integration-tests/src/router.rs

```diff
@@ -973,6 +974,530 @@ mod omnipool_router_tests {
+
+				let amount_to_sell = 20 * UNITS;
+				assert_ok!(Router::sell(
+					hydradx_runtime::RuntimeOrigin::signed(ALICE.into()),
+					insufficient_asset1,
+					insufficient_asset4,
+					amount_to_sell,
+					0,
+					trades
+				));
+
+				assert_balance!(ALICE.into(), HDX, 1000 * UNITS - 2 * ed);
+
+				TransactionOutcome::Commit(DispatchResult::Ok(()))
+			});
+		});
+	}
+
+	#[test]
+	fn ed_should_be_refunded_when_all_insufficient_assets_sold() {
+		TestNet::reset();
+
+		Hydra::execute_with(|| {
+			let _ = with_transaction(|| {
+				//Arrange
+				let name = b"INSUF1".to_vec();
+				let insufficient_asset1 = AssetRegistry::register_insufficient_asset(
+					None,
+					Some(name.try_into().unwrap()),
+					AssetKind::External,
+					Some(1_000),
+					None,
+					None,
+					None,
+					None,
+				)
+				.unwrap();
+
+				let name = b"INSUF2".to_vec();
+				let insufficient_asset2 = AssetRegistry::register_insufficient_asset(
+					None,
+					Some(name.try_into().unwrap()),
+					AssetKind::External,
+					Some(1_000),
+					None,
+					None,
+					None,
+					None,
+				)
+				.unwrap();
+
+				assert_ok!(Currencies::deposit(insufficient_asset1, &DAVE.into(), 100000 * UNITS,));
+				assert_ok!(Currencies::deposit(insufficient_asset2, &DAVE.into(), 100000 * UNITS,));
+				assert_ok!(Currencies::deposit(ETH, &DAVE.into(), 100000 * UNITS,));
+
+				assert_ok!(XYK::create_pool(
+					RuntimeOrigin::signed(DAVE.into()),
+					insufficient_asset1,
+					10000 * UNITS,
+					insufficient_asset2,
+					10000 * UNITS,
+				));
+
+				assert_ok!(XYK::create_pool(
+					RuntimeOrigin::signed(DAVE.into()),
+					insufficient_asset2,
+					10000 * UNITS,
+					ETH,
+					10000 * UNITS,
+				));
+
+				let trades = vec![
+					Trade {
+						pool: PoolType::XYK,
+						asset_in: insufficient_asset1,
+						asset_out: insufficient_asset2,
+					},
+					Trade {
+						pool: PoolType::XYK,
+						asset_in: insufficient_asset2,
+						asset_out: ETH,
+					},
+				];
+
+				let alice_balance_before_trade = Balances::free_balance(AccountId::from(ALICE));
+
+				let insufficient_asset1_balance = 100 * UNITS;
+				assert_ok!(Currencies::deposit(
+					insufficient_asset1,
+					&ALICE.into(),
+					insufficient_asset1_balance,
+				));
+
+				let extra_ed_charge = UNITS / 10;
+
+				let amount_to_sell = insufficient_asset1_balance;
+				assert_ok!(Router::sell(
+					hydradx_runtime::RuntimeOrigin::signed(ALICE.into()),
+					insufficient_asset1,
+					ETH,
+					amount_to_sell,
+					0,
+					trades
+				));
+				let alice_balance_after_trade = Balances::free_balance(AccountId::from(ALICE));
+
+				//ED should be refunded to alice as she sold all her asset, minus the 10% extra
+				assert_eq!(alice_balance_before_trade, alice_balance_after_trade + extra_ed_charge);
+
+				TransactionOutcome::Commit(DispatchResult::Ok(()))
+			});
+		});
+	}
+
+	#[test]
+	fn should_no_ed_charging_disabled_when_only_one_route_with_insufficient_assets() {
```

Comment:
I wouldn't know what this test is about just by this name

Reply:
ah yeah, fixed!

`ed_charging_should_not_be_disabled_when_only_one_trade_with_insufficient_assets`